File: python/Hashing/MD5.py
from hashlib import md5
"""
Module for calculating the MD5 hash of a string.

This module demonstrates how to use the standard `hashlib` library to generate the MD5 hash of a text.
MD5 is a widely used cryptographic hash algorithm, but it is no longer considered secure for cryptographic applications.
"""

import hashlib

def md5_hash(text: str) -> str:
	"""
	Computes the MD5 hash of a string.

	Args:
		text (str): The input string to hash.

	Returns:
		str: The MD5 hash in hexadecimal format.
	"""
	# Encode the string to bytes and compute the MD5 hash
	hash_object = hashlib.md5(text.encode('utf-8'))
	md5_hex = hash_object.hexdigest()
	return md5_hex


def main():
	"""
	More significant example usage of the md5_hash function.
	Demonstrates hashing multiple strings, comparing hashes, and hashing a file.
	"""
	# List of example texts
	texts = [
		"Hello, World!",
		"hello, world!",
		"The quick brown fox jumps over the lazy dog",
		"The quick brown fox jumps over the lazy dog."
	]
	print("MD5 hashes of different strings:")
	for t in texts:
		print(f"Text: {t}")
		print(f"MD5 Hash: {md5_hash(t)}\n")

	# Compare hashes of similar strings
	print("Comparing hashes of similar strings:")
	s1 = "password123"
	s2 = "password124"
	hash1 = md5_hash(s1)
	hash2 = md5_hash(s2)
	print(f"String 1: {s1} -> {hash1}")
	print(f"String 2: {s2} -> {hash2}")
	print(f"Are hashes equal? {'Yes' if hash1 == hash2 else 'No'}\n")


if __name__ == "__main__":
	main()

File: java_implementation/hashing/MD5.java
package java_implementation.hashing;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * Utility class for computing MD5 hashes.
 */
public class MD5 {
    /**
     * Computes the MD5 hash of a given string.
     *
     * @param input The input string to hash.
     * @return The MD5 hash as a hexadecimal string.
     */
    public static String md5Hash(String input) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] hashBytes = md.digest(input.getBytes());
            StringBuilder sb = new StringBuilder();
            for (byte b : hashBytes) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("MD5 algorithm not found", e);
        }
    }

    /**
     * Example usage of the md5Hash function.
     */
    public static void main(String[] args) {
        String[] texts = {
            "Hello, World!",
            "hello, world!",
            "The quick brown fox jumps over the lazy dog",
            "The quick brown fox jumps over the lazy dog."
        };
        System.out.println("MD5 hashes of different strings:");
        for (String text : texts) {
            System.out.println("Text: " + text);
            System.out.println("MD5 Hash: " + md5Hash(text) + "\n");
        }

        // Compare hashes of similar strings
        String s1 = "password123";
        String s2 = "password124";
        String hash1 = md5Hash(s1);
        String hash2 = md5Hash(s2);
        System.out.println("Comparing hashes of similar strings:");
        System.out.println("String 1: " + s1 + " -> " + hash1);
        System.out.println("String 2: " + s2 + " -> " + hash2);
        System.out.println("Are hashes equal? " + (hash1.equals(hash2) ? "Yes" : "No"));
    }
}
